File: 1387.sort-integers-by-the-power-value.cpp
/*
 * @lc app=leetcode id=1387 lang=cpp
 *
 * [1387] Sort Integers by The Power Value
 */
#include <bits/stdc++.h>

using std::unordered_map;
using std::vector;
using std::pair;

// @lc code=start
class Solution {
    int get_power(const int& val, unordered_map<int, int>& memo) {
        if (memo.find(val) != memo.end()) return memo[val];
        int steps = 0, cur = val;
        while (cur != 1) {
            if (cur & 1) {
                cur = cur * 3 + 1;
            } else {
                cur >>= 1;
            }
            ++steps;
        }
        memo[val] = steps;
        return steps;
    }

public:
    // T: O(n*log(n)), n := hi - lo
    // S: O(n)
    int getKth(int lo, int hi, int k) {
        vector<pair<int, int>> vec;
        vec.reserve(hi - lo + 1);
        unordered_map<int, int> memo;
        for (int x = lo; x <= hi; ++x) {
            vec.push_back({get_power(x, memo), x});
        }
        std::sort(vec.begin(), vec.end());
        return vec[k - 1].second;
    }
};
// @lc code=end

File: 1387.sort-integers-by-the-power-value.py
#
# @lc app=leetcode id=1387 lang=python3
#
# [1387] Sort Integers by The Power Value
#
import heapq

# @lc code=start
class Solution:
    # Memoized recursion + heap / sorting
    # T: O(n * log(k)), n := hi - lo
    # S: O(n) recursion
    def getKth(self, lo: int, hi: int, k: int) -> int:
        memo = dict()
        memo[1] = 0

        def calc_steps(x: int) -> int:
            if x not in memo:
                memo[x] = 1 + calc_steps(x * 3 + 1 if x & 1 else x >> 1)
            return memo[x]

        pq = []
        for x in range(lo, hi + 1):
            # To keep the heap complexity low, use a max heap and 
            # pop the larger values on the way, so that the heap 
            # top ends up with the k-th element
            heapq.heappush(pq, (-calc_steps(x), -x))
            if len(pq) > k:
                heapq.heappop(pq)
        return -pq[0][1]
        
# @lc code=end

